DIPs/DIP1xxx-rval_to_ref.md

```diff
@@ -0,0 +1,283 @@
+
+That said, D has the `out` attribute, which is a semantic statement of this intent. It could be that this proposal is amended to include non-const ref arguments, expecting that `out` shall be used exclusively to mark this intent.
+If we assume that world, and `out` is deployed appropriately, there are 2 cases where mutable-ref may be used:
+ 1. When the function *modifies* the input; not a strict output parameter, but still outputs new information
+ 2. Still used as input, but a user is trying to subvert the restrictiveness of D's `const`
+
+The proposal could be amended to accept mutable ref's depending on the value-judgement balancing these 2 use cases.
+Sticking with `const` requires no such value judgement to be made at this time, and it's much easier to relax the spec in the future with emergence of evidence to do so.
+
+## Why not `auto ref`?
+
+A frequently proposed solution to this situation is to receive the arg via `auto ref`.
+
+`auto ref` solves a different set of problems; those may include "pass this argument in the most efficient way", or "forward this argument exactly how I received it". The implementation of `auto ref` requires that every function also be a template.
+
+There are many reasons why every function can't or shouldn't be a template.
+1. API is not your code, and it's not already `auto ref`
+2. Is distributed as a binary lib
+3. Is exported from DLL
+4. Is virtual
+5. Is extern(C++)
+6. Intent to capture function pointers or delegates
+7. Has many args; unreasonable combinatorial explosion
+8. Is larger-than-inline scale; engineer assesses that reducing instantiation bloat has greater priority than maximising parameter passing efficiency in some cases
+
+Any (or many) of these reasons may apply, eliminating `auto ref` from the solution space.
+
+## Key use cases
+
+By comparison, C++ has a very high prevalence of `const&` args and classes with virtual functions, and when interfacing with C++, those functions are mirrored to D. The issue addressed in this DIP becomes magnified significantly to this set of users.
+
+The D community has invested significant resources in improving interaction with C++; either co-existing or simplifying a migration, and thereby make D attractive to the C++ audience.
+The importance of this initiative is widely agreed; it has featured prominently in the bi-annual game-plans documents, and comprehensive interaction with even the C++ standard library has attracted funding from the D foundation.
+This DIP offers a lot for interaction with C++ APIs.
+
+This issue is also likely to appear more frequently for vendors with tight ABI requirements.
+Users of closed-source libraries distributed as binary libs, or libraries distributes as DLLs are more likely to encounter these challenges interacting with those APIs as well.
+
+Another high-probability occurrence is OOP, where virtual function APIs inhibit the use of templates.
```

Comment:
Another good use case is random generation. You always should take the generator by `ref` to avoid accidentally copying the generator (oops), but that makes the API more verbose as the user can't declare the RandomEngine and use it one line.
Phobos uses `auto ref`- it shouldn't :/
A summary of Phobos's broken std.random: https://dconf.org/2015/talks/wakeling.html